### ISSUE TITLE
Feature Request - hoverOpts - control units

### ISSUE THREAD

**Issue by jvp-gama** [rstudio/shiny#2079]
I would like to request that an additional parameter be added to hoverOpts and plot interaction control functions interacting with plot_output and image_output. This parameter will control the units assigned to the x y values returned by input$plot_hover, input$plot_click etc values. Currently this is assumed based on the output function plot_output or image_output. with plot_output, the units are in the same units as the data, and with image_output the units are in pixels and based on location of the mouse.

I am using treemap data and I need to associate the mouse location with one of the boxes in the treemap. treemap software will calculate the coordinates of the boxes, however, when I plot a treemap using plot_output, i receive x/y coordinates in terms of the data and this is not useful. I would like to be able to define the units as "data" or "pixel" as a parameter of the hoverOpts function. This way I can get pixel location using the plot_output function.

I then can use the pixel coordinate together with the coordinates the treemap software calculates to identify which box the user has selected. With this I can then do complex follow up calculation for the user based on the box selected in the plot.

something like 
hoverOpts(id = NULL, delay = 300, delayType = c("debounce", "throttle"),
  clip = TRUE, nullOutside = TRUE, unitsType = c("data","pixels"))

**Comment by jvp-gama** [comment]
Thank you, very helpful.  I learned that I can access functions not exposed.  Good to know.  

I took your example and expanded so that you can see the solution to my problem.

```library(shiny)
library(treemap)
library(treemapify)
library(dplyr)
library(ggplot2)
# A demonstration of clicking, hovering, and brushing
shinyApp(
  ui = basicPage(
    fluidRow(
      column(width = 5,
             plotOutput("plot", height=300,
                        click = "plot_click",
                        hover = hoverOpts(
                          id = "plot_hover",
                          delay = 200,
                          delayType = "throttle"
                        )
             )
      ),
      column(width = 4,
             verbatimTextOutput("plot_clickinfo"),
             verbatimTextOutput("plot_hoverinfo")
      )
    ),
    fluidRow(
      column(width = 5,
             plotOutput("heatmap", width = "800px", height = "800px",
                        click = "plot_click_heat",
                        hover = hoverOpts(
                          id = "plot_hover_heat",
                          delay = 200,
                          delayType = "throttle"
                        )
             )
      ),
      column(width = 4,
             verbatimTextOutput("plot_clickinfo_heat"),
             verbatimTextOutput("plot_hoverinfo_heat"),
             tableOutput("point")
      )
    )
  ),
  server = function(input, output, session) {
    output$plot <- renderPlot({
      plot(cars$speed, cars$dist)
    })
    output$heatmap <- renderPlot({
      data(GNI2014)
    tm <- ggplot(GNI2014,aes(area = population, fill = population, subgroup = continent,  label = iso3)) +
      geom_treemap()+
      geom_treemap_subgroup_border(colour = "black") +
      geom_treemap_subgroup_text(place = "centre", grow = F, alpha = 0.5, colour =
                                   "black", fontface = "italic", min.size = 0) +
      geom_treemap_text( reflow = TRUE, color = "grey70",place = "topleft") +
      scale_fill_gradient2(low = "firebrick3", mid = "lightyellow", high = "green3", midpoint=0)
    tm
    })
    
    output$plot_clickinfo <- renderPrint({
      cat("Click:\n")
      str(shiny:::scaleCoords(input$plot_click$x, input$plot_click$y, input$plot_click))
      str(input$plot_click)
    })
    output$plot_hoverinfo <- renderPrint({
      cat("Hover:\n")
      str(shiny:::scaleCoords(input$plot_hover$x, input$plot_hover$y, input$plot_hover))
    })
    output$plot_clickinfo_heat <- renderPrint({
      cat("Click:\n")
      str(shiny:::scaleCoords(input$plot_click_heat$x, input$plot_click_heat$y, input$plot_click_heat))
      str(input$plot_click_heat)
    })
    output$plot_hoverinfo_heat <- renderPrint({
      cat("Hover:\n")
      str(shiny:::scaleCoords(input$plot_hover_heat$x, input$plot_hover_heat$y, input$plot_hover_heat))
    })
    output$point <- renderTable({
      if (is.null(input$plot_click_heat)) return()
      hover <- input$plot_click_heat
      c <- shiny:::scaleCoords(hover$x, hover$y, hover)
      c$x <- (c$x / (hover$range$right+1))
      c$y <- 1- (c$y / (hover$range$bottom+1))
      data(GNI2014)
      df <- treemapify(GNI2014,  area = "population",
                       subgroup = "continent",subgroup2 = "iso3")
      df$continent <- as.character(df$continent)
      
      point <- df %>%
        dplyr::filter(xmin < c$x) %>%
        dplyr::filter(xmax > c$x) %>%
        dplyr::filter(ymin < c$y) %>%
        dplyr::filter(ymax > c$y)
      
      if (nrow(point) == 0) return()
      point
      
    })
  }
)
```

**Comment by wch** [comment]
Great. I'm reopening because I think this is a useful feature request -- using non-exported functions is generally not a good idea.

**Comment by schloerke** [comment]
resolved in #2183
